src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2623,157 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+
+        // If overridden Equals or GetHashCode found, stop searching further.
+        return TRUE;
+    }
+
+    // The type itself did not override Equals or GetHashCode, go for its fields.
+    ApproxFieldDescIterator iter = ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
+    FieldDesc* pField = iter.Next();
+    while (pField != NULL
+        && pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+        && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
+    {
+        // We cannot find overridden method in non-class types.
+        pField = iter.Next();
+    }
+
+    if (pField == NULL)
+    {
+        // No more instance fields. Now we can say this type did not override Equals or GetHashCode.
+        mt->SetHasCheckedOverriddenEqualsOrGetHashCode();
+        return FALSE;
+    }
+
+    BOOL hasOverriddenEqualsOrGetHashCode = FALSE;
+
+    do
+    {
+        // Try to find the overridden methods among all instance fields recursively.
+        MethodTable* fieldMethodTable = pField->LookupFieldTypeHandle().GetMethodTable();
+        hasOverriddenEqualsOrGetHashCode = HasOverriddenEqualsOrGetHashCodeMethodWorker(fieldMethodTable);
+        pField = iter.Next();
+    } while(!hasOverriddenEqualsOrGetHashCode && pField != NULL);
+
+    // We've gone through all instance fields. It's time to cache the result.
+    mt->SetHasCheckedOverriddenEqualsOrGetHashCode();
+    if (hasOverriddenEqualsOrGetHashCode)
+    {
+        mt->SetHasOverriddenEqualsOrGetHashCode();
+    }
+
+    return hasOverriddenEqualsOrGetHashCode;
+}
+
+static BOOL HasOverriddenEqualsOrGetHashCodeMethod(MethodTable* mt)
+{
+    STANDARD_VM_CONTRACT;
+
+    if (mt->HasCheckedOverriddenEqualsOrGetHashCode())
+    {
+        return mt->HasOverriddenEqualsOrGetHashCode();
+    }
+
+    if (g_pValueTypeMT == NULL)
+    {
+        g_pValueTypeMT = MscorlibBinder::GetClass(CLASS__VALUE_TYPE);
+    }
+
+    if (g_slotEquals == 0)
+    {
+        g_slotEquals = MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__EQUALS)->GetSlot();
+    }
+
+    if (g_slotGetHashCode == 0)
+    {
+        g_slotGetHashCode = MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__GET_HASH_CODE)->GetSlot();
+    }
+
+    BOOL hasOverriddenMethod = HasOverriddenEqualsOrGetHashCodeMethodWorker(mt);
+
+    return hasOverriddenMethod;
+}
+
+// Return true if the valuetype does not contain pointer, is tightly packed, 
+// does not have floating point number field and does not override Equals method.
 FCIMPL1(FC_BOOL_RET, ValueTypeHelper::CanCompareBits, Object* obj)
 {
     FCALL_CONTRACT;
 
     _ASSERTE(obj != NULL);
     MethodTable* mt = obj->GetMethodTable();
-    FC_RETURN_BOOL(!mt->ContainsPointers() && !mt->IsNotTightlyPacked());
+    if (mt->ContainsPointers()
+        || mt->IsNotTightlyPacked()
+        || mt->HasFloatingPointNumberField())
+    {
+        FC_RETURN_BOOL(FALSE);
+    }
+
+    OBJECTREF objRef = ObjectToOBJECTREF(obj);
```

Comment:
Could you please measure how much slower is the "fast equals" now? I think this will need to be structured differently to avoid the regression.